### PR TITLE
GCW-3665 Set offer shareable expires_at on offer close

### PR DESCRIPTION
**Pull request by swatijadhav**
### Ticket Link: 
https://jira.crossroads.org.hk/browse/GCW-3665

### What does this PR do?
Set offer shareable expires_at on offer close

### CODE REVIEW

File: app/models/offer.rb [hunk 1]
```diff
@@ -191,6 +191,12 @@ def set_initial_state
       offer.inactive_at = Time.now
     end
 
+    after_transition on: :cancel do |offer, _transition|
+      if User.current_user == offer.created_by
+        offer.expire_shareable_resource
+      end
```
Comment (patrixr) [inline app/models/offer.rb:197]: why the user condition 🤔  ?

File: app/models/offer.rb [hunk 2]
```diff
@@ -259,6 +265,16 @@ def offers_count_per_state(offer)
     end
   end
 
+  def expire_shareable_resource(expires_at=Time.now)
+    return if self.shareable.blank?
+    expires_at = DateTime.strptime(expires_at, '%d/%m/%y %H:%M %z') if expires_at.is_a?(String)
```
Comment (patrixr) [inline app/models/offer.rb:270]:
```suggestion
    expires_at = Time.parse(expires_at) if expires_at.is_a?(String)
```
(we don't want to assume the format, that's a ticking time bomb)

File: app/models/offer.rb [hunk 3]
```diff
@@ -259,6 +265,16 @@ def offers_count_per_state(offer)
     end
   end
 
+  def expire_shareable_resource(expires_at=Time.now)
```
Comment (patrixr) [inline app/models/offer.rb:268]: This is a useful method, but I don't think it should be on the offer. Plus, we'll probably need it for other models in the future.

Can we move it to a static Shareable method ?

e.g

```ruby
Shareable.expire(offer, time)
Shareable.expire(offer.packages, time)
```

**Review by patrixr — COMMENTED** [review]
Looks okay overall, just dropped a couple of minor comments

File: app/models/offer.rb (hunk 1 above)
Comment (swatijadhav) [reply app/models/offer.rb:197]: When Donor cancels offer, we want to auto stop sharing that offer and hence condition added.

Comment (patrixr) [reply app/models/offer.rb:197]: My point is that I think we should cancel regardless of who cancels the offer

Comment (patrixr) [reply app/models/offer.rb:197]: Also, it is dangerous to rely on a state (current_user) which only exists in the context of a request.

**Review by patrixr — APPROVED** [review]

**Pull request merged**